File: src/algorithms/dynamic-programming/max-subset-sum-no-adjacent/solution2.js
// O(n) time | O(1) space
function maxSubsetSumNoAdjacent(array) {
  if (!array.length) return 0;
  if (array.length === 1) return array[0];
  let second = array[0];
  let first = Math.max(array[0], array[1]);
  for (let i = 2; i < array.length; i++) {
  	const current = Math.max(first, second + array[i]);
  	second = first;
  	first = current;
  }
  return first;
}

exports.maxSubsetSumNoAdjacent = maxSubsetSumNoAdjacent;

File: src/algorithms/dynamic-programming/max-subset-sum-no-adjacent/solution2.go
package main

// O(n) time | O(1) space
func MaxSubsetSumNoAdjacent(array []int) int {
	if len(array) == 0 {
		return 0
	} else if len(array) == 1 {
		return array[0]
	}
	first := max(array[0], array[1])
	second := array[0]
	for i := 2; i < len(array); i++ {
		current := max(first, second+array[i])
		second = first
		first = current
	}
	return first
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}